feat(issueless events) Skip group creation for transaction events

fpacifici opened the pull request:
This is the step that actually removes the event creation for transaction events.
Ideally there should be a bigger refactoring of event_manager to have two separate save path for errors and transactions, but this is hard to obtain without a bigger refactoring of EventManager.

Still WIP since I want to add a feature to gate the issueless event ingestion.

fpacifici commented on src/sentry/event_manager.py:
```diff
@@ -787,39 +811,44 @@ def save(self, project_id, raw=False, assume_normalized=False):
+                        },
+                    })
+                )
+        if frequencies:
```
We may want to add some frequencies stats for issueless events.

markstory replied:
Would we need a way to identify different transactions to do that?

jan-auer commented:
Opinion: You can probably get away without a feature flag. If people want to start ingesting a payload, let them. The only thing we need to make sure is it doesn't break ingest.

fpacifici commented:
@jan-auer (besides data integrity) my concern is people ingesting issueless event and some other obscure feature breaking when it tries to load an event starting from a project and id and not finding a group.

tkaemming commented on src/sentry/event_manager.py:
```diff
@@ -566,6 +566,28 @@ def save(self, project_id, raw=False, assume_normalized=False):
+        if event.get_event_type() == 'transaction':
+            create_group = False
+        else:
+            create_group = True
```
nit: This doesn't mean that the event /will/ create a group, just that it /can/ create a group. Maybe this should be called `can_create_group` or something? That's not a great name either, but I think something more descriptive would help anybody who isn't familiar with this method from getting tripped up while reading it.

fpacifici replied:
renamed to issueless_event and inverted the condition

tkaemming commented on src/sentry/event_manager.py:
```diff
@@ -566,6 +566,28 @@ def save(self, project_id, raw=False, assume_normalized=False):
+
+        org = Organization.objects.get_from_cache(
+            id=project.organization_id
+        )
```
Isn't this already bound on [line 506](https://github.com/getsentry/sentry/blob/148a0c22f5f6fbea50f660cab78f93ebcec215a9/src/sentry/event_manager.py#L506-L507)?

fpacifici replied:
correct. will fix

tkaemming commented on src/sentry/event_manager.py:
```diff
@@ -566,6 +566,28 @@ def save(self, project_id, raw=False, assume_normalized=False):
+                    'event_uuid': data['event_id'],
+                    'project_id': project.id,
+                }
+            )
```
FYI @JTCunning.

JTCunning replied:
This is fine as of now. If anything, it will alert us as to people who are possibly doing things they shouldn't.

fpacifici commented:
Test plan:
- added tests for ingestion of issueless event
- added test for event detail api to ensure issueless event are properly loaded
- Ingesting standard event through the python api still work:
-- events with group https://gist.github.com/fpacifici/cb7e80c7bf36d5a9ebbd3950f5b9ee52
-- the events and the groups are visible in the UI
- Ingesting a transaction event without having the feature enable fails and logged
-- command https://gist.github.com/fpacifici/7dbb0c2c83c5b38df2255b58b639c5d5
```
00:12:22 worker  | 00:12:22 [INFO] sentry.events: issueless.disabled (project_id=2L event_uuid=u'c03dfbab8a8145eeaa0d1a1adfcfcae0')

c0808f9266f3 :) select event_id, group_id, project_id, message, type from sentry_local where project_id = 2 and event_id = 'c03dfbab8a8145eeaa0d1a1adfcfcae0';

SELECT 
    event_id, 
    group_id, 
    project_id, 
    message, 
    type
FROM sentry_local 
WHERE (project_id = 2) AND (event_id = 'c03dfbab8a8145eeaa0d1a1adfcfcae0')

Ok.

0 rows in set. Elapsed: 0.008 sec. 

```

- Ingesting an issueless event works when enabling the feature:
```
c0808f9266f3 :) select event_id, group_id, project_id, message, type from sentry_local where project_id = 2 and event_id = 'c03dfbab8a8145eeaa0d1a1adfcfcae0';

SELECT 
    event_id, 
    group_id, 
    project_id, 
    message, 
    type
FROM sentry_local 
WHERE (project_id = 2) AND (event_id = 'c03dfbab8a8145eeaa0d1a1adfcfcae0')

┌─event_id─────────────────────────┬─group_id─┬─project_id─┬─message─┬─type────────┐
│ c03dfbab8a8145eeaa0d1a1adfcfcae0 │        0 │          2 │ wait    │ transaction │
└──────────────────────────────────┴──────────┴────────────┴─────────┴─────────────┘

1 rows in set. Elapsed: 0.009 sec. 

c0808f9266f3 :) 
```

response from event details api for an issueless event: https://gist.github.com/fpacifici/08c3183f2825b00cab282b2bf2e65774

fpacifici commented:
Removed the feature gating as per offline discussion and rebased on master so this does not need rebase anymore.

@tkaemming, regarding the hashes and fingerprint, I tried to reduce the change as much as possible. I believe the grouping code can be safely skipped here but I could not confirm it.

untitaker commented on tests/snuba/api/endpoints/test_event_details.py:
```diff
@@ -330,3 +330,42 @@ def test_event_ordering(self):
+            'type': 'transaction',
+        }
+
+        with self.feature('organizations:issueless-events'):
```
This should not be necessary anymore

fpacifici replied:
removed

fpacifici commented:
@dcramer rule, engine, service hooks and data forwarding has already been removed for issueless events in #13982
